### ISSUE TITLE
Optimised TLR release

### ISSUE THREAD

**Issue by chrisvest** [chrisvest/stormpot#79]
In BlazePool, `Slot.release()` currently has to do a CAS because the transitions TLR_CLAIMED->LIVE and TLR_CLAIMED->CLAIMED can race.

The latter transition happen when a contending thread removes the object from the live-queue. A characteristic of a tlr-claimed object is that it is simultaneously claimed and in the live-queue, while ordinarily claimed objects are not in the live-queue. Therefore, when a thread removes an object from the live-queue, and can't claim it because it is already claimed, then it must transition it from TLR_CLAIMED to CLAIM, since it's now no longer also in the live-queue. This is important because 1) a release from CLAIMED adds it back to the live-queue, but a release from TLR_CLAIMED does not since it would leak memory by putting duplicates into the live-queue, and 2) the live-queue must eventually end up empty if there are no objects available for claiming, since that is where the blocking behaviour comes from.

This whole thing can be optimised. If the TLR_CLAIMED->CLAIMED transition was removed, then there would be no racing transitions *out of* CLAIMED or TLR_CLAIMED, and `Slot.release()` could therefore be implemented with the much faster `putOrderedInt`. This might be possible if the following modification was made to the slow-claim-path: 1) if a TLR_CLAIMED slot is removed from the live-queue, *it is not* transitioned to any other state, but instead CAS'ed onto a spill-stack. The absence of this transition enables the `putOrderedInt` optimisation. 2) Upon entry to the slow-claim-path, the entire spill-stack is emptied back into the live-queue. This ensures that the live-queue will eventually become empty if the pool is depleted. 3) The blocking wait on the queue is taken in small quantums, where in between the spill-stack is emptied back into the live-queue. This ensures that the blocking threads will regularly recheck if the slots in the spill-stack are still TLR_CLAIMED.

A drawback to this approach is that threads in a system under high load with a highly contended pool, will most likely be extremely unfair to the threads that end up blocking, letting the lucky TLR_CLAIMING threads have all the fun.

**Comment by chrisvest** [comment]
This has been merged into master. Docs and comments might still need revision, though.
